modules/api/src/main/java/org/apache/ignite/lang/IgniteExceptionMapper.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.lang;
+
+import java.util.function.Function;
+
+/**
+ * Represents a mapper from an internal exception {@code T} to a public one {@code R}.
+ */
+public class IgniteExceptionMapper<T extends Exception, R extends Exception> {
+    /** Class that represents an internal exception for mapping. */
+    private final Class<T> from;
+
+    /** Mapping function. */
+    private final Function<T, R> mapper;
+
+    /**
+     * Creates a new instance of mapper.
+     *
+     * @param from Class instance that represents a class of internal exception.
+     * @param mapper Mapping function to map an internal exception to a public one.
+     */
+    private IgniteExceptionMapper(Class<T> from, Function<T, R> mapper) {
+        this.from = from;
+        this.mapper = mapper;
+    }
+
+    /**
+     * Returns a class instance that represents an internal exception to be used for mapping.
+     *
+     * @return Class instance that represents an internal exception to be used for mapping.
+     */
+    public Class<T> mappingFrom() {
+        return from;
+    }
+
+    /**
+     * Maps the provided internal exception to a public one.
+     *
+     * @param error Exception instance to be mapped.
+     * @return Public exception instance.
+     */
+    public R map(T error) {
+        return mapper.apply(error);
+    }
+
+    /**
+     * Creates a new exception mapper from an internal exception {@code T} to a public runtime exception {@code R}.
+     *
+     * @param from Class instance that represents a class of internal exception.
+     * @param mapper Mapping function to map an internal exception to a public one.
+     * @param <T> Internal exception type.
+     * @param <R> Public runtime exception type.
+     *
+     * @return New instance of {@link IgniteExceptionMapper}.
+     */
+    public static <T extends Exception, R extends IgniteException> IgniteExceptionMapper<T, R> unchecked(
```

Comment:
The main goal is to restrict the type of mapper. The exception mapper should provide either a mapping Exception -> IgniteException (unchecked) or Exception -> IgniteCheckedException (checked). The mapper should not try to provide a mapping Exception -> IgniteInternalException, for instance.
Java generics do not allow using something as follows:
```
    <T, R extends IgniteException> Function<T, R> mapper();
    <T, R extends IgniteCheckedException> Function<T, R> mapper();
```
just because these methods have exactly the same `erasure`

Reply:
Yep, ok.